Add KeyVault Certificate Functions

Billy-Sheppard opened the pull request:
Apologies if this was already worked on or out of spec, adapted directly from `secret.rs`, welcome any improvements.

- moved `enum RecoveryLevel` to the key_vault `lib.rs`
- added `certificate.rs` based directly off `secret.rs`
- potentially `fn _update_certificate_policy()` is not needed/required
- documentation may not be suitable

heaths commented on sdk/key_vault/src/certificate.rs:
````diff
@@ -0,0 +1,816 @@
+    ///
+    /// Runtime::new().unwrap().block_on(example());
+    /// ```
+    pub async fn delete_certificate(&mut self, certificate_name: &'a str) -> Result<(), Error> {
````
This will actually need to return a long-running operation, but as with pageables, I think can be in a separate PR we can do later. I'll open an issue and would recommend adding a comment to reference the issue as a TODO.

heaths replied:
We can use https://github.com/Azure/azure-sdk-for-rust/issues/174 for tracking.

heaths reviewed (changes requested):
Overall, the implementation looks good. Most of the requested changes are to align with our other SDKs, since we consistency after idiomacy (and naming - apart from casing - is rarely idiomatic). The biggest change, though, is to define a `CertificateProperties` and use that in various structs and methods, which actually makes calling the code much easier and tested well in UX studies. This is fairly consistent across our SDKs - even beyond Key Vault.

Billy-Sheppard commented:
Thanks for your comments Heath, I (think) I've addressed all but the `CertificateClient` issue - do you have a rust code example I can build off, otherwise I'll decipher from the MS docs.

Let me know if anything else needs changing, potentially (probably) there are some cleaner ways to address some of the fixes I've implemented.

Thanks again.

Billy-Sheppard commented:
Apologies for the messy history, I think I have resolved all the changes requested but I am unsure on how to mark them properly? Is marking comments as resolved incorrect?

Cheers

Billy-Sheppard commented:
Apologies for the messy history, it took me a while to understand the ins and outs of the code, especially the tests, sorry, but I think I have resolved all the changes requested. 

However, I am unsure on how to mark them properly? Is marking comments as resolved incorrect?

Cheers

Billy-Sheppard commented:
@rylev 

Hi, I'm pretty keen to get this reviewed, as we're currently using a fork in production, I understand @heaths is busy with a bunch of projects, are you (Ryan) able to review this?

Thanks

heaths commented:
Could you resolve the conflicts and I'll take a look? I expect a few significant changes.

Billy-Sheppard commented:
@heaths - are you able to complete this/provide further feedback? Do the remaining tasks require any further improvement? I also opened issue: https://github.com/Azure/azure-sdk-for-rust/issues/516.

Thanks

heaths commented on sdk/key_vault/src/client.rs:
````diff
@@ -164,6 +164,164 @@ impl<'a, T: TokenCredential> KeyClient<'a, T> {
+    /// let creds = DefaultCredential::default();
+    /// let client = CertificateClient::new("test-key-vault.vault.azure.net", &creds).unwrap();
+    /// ```
+    pub fn new(vault_url: &str, token_credential: &'a T) -> Result<Self, Error> {
````
Issue #516 opened.

heaths reviewed (approved):
Thank you.
